### PR TITLE
Feature/doctor approve appointment

### PR DESCRIPTION
**Pull request by mohamedebrahim2020**

### CODE REVIEW

File: app/Http/Controllers/AppointmentController.php [hunk 1]
```diff
@@ -18,6 +17,17 @@ public function __construct(AppointmentService $service)
...
+    public function index()
+    {
+        $appointments = $this->service->index();
+        return response()->json(IndexAppointmentResource::collection($appointments), Response::HTTP_OK);
+    }
+
+    public function approve()
+    {
+        $this->service->approve();
+        return response()->json([], Response::HTTP_OK);
```
Comment (MohannadElemary2) [inline app/Http/Controllers/AppointmentController.php]: Send the appointment ID from the controller to the service. The service should not get any data directly from the request so it become reusable.

File: app/Repositories/AppointmentRepository.php [hunk 2]
```diff
@@ -18,6 +17,11 @@ public function __construct(Appointment $model)
        parent::__construct($model);
    }
 
+   public function filterAppointmentsByStatus()
+   {
+       return request()->user('doctor')->appointments()->filter(app(AppointmentFilters::class))->get();
```
Comment (MohannadElemary2) [inline app/Repositories/AppointmentRepository.php]: use `auth()` facade to access current user

File: app/Services/AppointmentService.php [hunk 3]
```diff
@@ -13,6 +15,37 @@ public function __construct(AppointmentRepository $repository)
...
+        $doctor = request()->user();
+        $this->checkDoctorHasThisAppointment($doctor, $appointment);
+        $this->checkAvailabiltyToApprove($appointment);
+        $this->update(['status' => AppointmentStatus::APPROVED] , $appointment->id);
+        $appointment->patient->notify(new AppointmentNotification($this->repository->find(request()->appointment)));
+    }
+
+    public function checkDoctorHasThisAppointment($doctor, $appointment)
+    {
+        $doctor->appointments->contains($appointment) ? "" : abort(Response::HTTP_METHOD_NOT_ALLOWED);
```
Comment (MohannadElemary2) [inline app/Services/AppointmentService.php]: `HTTP_METHOD_NOT_ALLOWED` is not used in this context. Please use `BAD REQUEST`

File: app/Services/AppointmentService.php [hunk 4]
```diff
@@ -13,6 +15,37 @@ public function __construct(AppointmentRepository $repository)
...
+    }
+
+    public function checkAvailabiltyToApprove($appointment)
+    {
+        $nowInMs = Carbon::now()->timestamp * 1000;
+        if ($appointment->status ==2 || $appointment->time <= $nowInMs) {
```
Comment (MohannadElemary2) [inline app/Services/AppointmentService.php]: Use `ENUM` for status

File: app/Http/Controllers/AppointmentController.php (hunk 1 above)
Comment (mohamedebrahim2020) [reply app/Http/Controllers/AppointmentController.php]: done

File: app/Repositories/AppointmentRepository.php (hunk 2 above)
Comment (mohamedebrahim2020) [reply app/Repositories/AppointmentRepository.php]: done

File: app/Services/AppointmentService.php (hunk 3 above)
Comment (mohamedebrahim2020) [reply app/Services/AppointmentService.php]: done

File: app/Services/AppointmentService.php (hunk 4 above)
Comment (mohamedebrahim2020) [reply app/Services/AppointmentService.php]: done

File: app/Http/Controllers/AppointmentController.php [hunk 5]
```diff
@@ -18,6 +18,17 @@ public function __construct(AppointmentService $service)
         $this->service = $service;
     }
 
+    public function index()
+    {
+        $appointments = $this->service->index();
+        return response()->json(IndexAppointmentResource::collection($appointments), Response::HTTP_OK);
+    }
+
+    public function approve()
```
Comment (MohannadElemary2) [inline app/Http/Controllers/AppointmentController.php]: Send the id to the service from here. The service shouldn't access the request directly
```
public function approve($id){
        $appointment = $this->service->approve($id);
        ....
}
```

File: app/Repositories/AppointmentRepository.php (hunk 2 above)
Comment (mohamedebrahim2020) [reply app/Repositories/AppointmentRepository.php]: done

File: app/Http/Controllers/AppointmentController.php (hunk 5 above)
Comment (mohamedebrahim2020) [reply app/Http/Controllers/AppointmentController.php]: done

**Pull request merged**